GPU/GPUTracking/Base/GPUParam.inc

```diff
@@ -0,0 +1,136 @@
+{
+  // conversion of coorinates global->sector
+  *x = X * SliceParam[iSlice].CosAlpha + Y * SliceParam[iSlice].SinAlpha;
+  *y = Y * SliceParam[iSlice].CosAlpha - X * SliceParam[iSlice].SinAlpha;
+  *z = Z;
+}
+
+MEM_CLASS_PRE()
+GPUdi() float MEM_LG(GPUParam)::GetClusterRMS(int yz, int type, float z, float angle2) const
+{
+  //* recalculate the cluster error wih respect to the track slope
+
+  MakeType(const float*) c = ParamRMS0[yz][type];
+  float v = c[0] + c[1] * z + c[2] * angle2;
+  v = CAMath::Abs(v);
+  return v;
+}
+
+MEM_CLASS_PRE()
+GPUdi() void MEM_LG(GPUParam)::GetClusterRMS2(int iRow, float z, float sinPhi, float DzDs, float& ErrY2, float& ErrZ2) const
+{
+  int rowType = tpcGeometry.GetROC(iRow);
+  if (rowType > 2) {
+    rowType = 2; // TODO: Add type 3
+  }
+  z = CAMath::Abs((250.f - 0.275f) - CAMath::Abs(z));
+  float s2 = sinPhi * sinPhi;
+  if (s2 > 0.95f * 0.95f) {
+    s2 = 0.95f * 0.95f;
+  }
+  float sec2 = 1.f / (1.f - s2);
+  float angleY2 = s2 * sec2;          // dy/dx
+  float angleZ2 = DzDs * DzDs * sec2; // dz/dx
+
+  ErrY2 = GetClusterRMS(0, rowType, z, angleY2);
+  ErrZ2 = GetClusterRMS(1, rowType, z, angleZ2);
+  ErrY2 *= ErrY2;
+  ErrZ2 *= ErrZ2;
+}
+
+MEM_CLASS_PRE()
+GPUdi() float MEM_LG(GPUParam)::GetClusterError2(int yz, int type, float z, float angle2) const
+{
+  //* recalculate the cluster error wih respect to the track slope
+
+  MakeType(const float*) c = ParamS0Par[yz][type];
+  float v = c[0] + c[1] * z + c[2] * angle2 + c[3] * z * z + c[4] * angle2 * angle2 + c[5] * z * angle2;
+  v = CAMath::Abs(v);
+  if (v < 0.01f) {
+    v = 0.01f;
+  }
+  v *= yz ? rec.ClusterError2CorrectionZ : rec.ClusterError2CorrectionY;
+  return v;
+}
+
+MEM_CLASS_PRE()
+GPUdi() void MEM_LG(GPUParam)::GetClusterErrors2(int iRow, float z, float sinPhi, float DzDs, float& ErrY2, float& ErrZ2) const
+{
+  // Calibrated cluster error from OCDB for Y and Z
+  int rowType = tpcGeometry.GetROC(iRow);
+  if (rowType > 2) {
+    rowType = 2; // TODO: Add type 3
+  }
+  z = CAMath::Abs((250.f - 0.275f) - CAMath::Abs(z));
+  float s2 = sinPhi * sinPhi;
+  if (s2 > 0.95f * 0.95f) {
+    s2 = 0.95f * 0.95f;
+  }
+  float sec2 = 1.f / (1.f - s2);
+  float angleY2 = s2 * sec2;          // dy/dx
+  float angleZ2 = DzDs * DzDs * sec2; // dz/dx
+
+  ErrY2 = GetClusterError2(0, rowType, z, angleY2);
+  ErrZ2 = GetClusterError2(1, rowType, z, angleZ2);
+}
+
+MEM_CLASS_PRE()
+GPUdi() void MEM_LG(GPUParam)::UpdateClusterError2ByState(short clusterState, float& ErrY2, float& ErrZ2) const
+{
+  if (clusterState & GPUTPCGMMergedTrackHit::flagEdge) {
+    ErrY2 += 0.35f;
+    ErrZ2 += 0.15f;
+  }
+  if (clusterState & GPUTPCGMMergedTrackHit::flagSingle) {
+    ErrY2 += 0.2f;
+    ErrZ2 += 0.2f;
+  }
+  if (clusterState & (GPUTPCGMMergedTrackHit::flagSplitPad | GPUTPCGMMergedTrackHit::flagShared | GPUTPCGMMergedTrackHit::flagSingle)) {
```

Comment:
You have the same check on sharing in two non-exclusive comparisons, so the errors will be augmented  twice, need to add else?

Reply:
@shahor02 : The checks are indeed correct this way. the shared flag is checked twice, because it will increase the errors in y and in z. And the single flag is checked three time, since it shall increase both errors even further. One could have seperate flags fir single pad / single time, but that would require one more bit in the cluster data format as well, so I didn't do that.

Reply:
Indeed, sorry, did not notice when checking on the phone that different ``if's`` act on different errors.